Remove all uses of Tag from climada/engine

peanutfun opened the pull request:
Changes proposed in this PR:
- Remove all uses of `Tag` from `climada/engine/`

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]
- [x] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

chahank commented on climada/engine/impact.py:
```diff
@@ -1215,20 +1171,6 @@ def from_hdf5(cls, file_path: Union[str, Path]):
             ├─ event_name
             ├─ frequency
             ├─ imp_mat
```
I think the `haz_type` is now missing from the description docstring. Should be in the `.attrs` ?

peanutfun replied:
True, looks like we are not writing it anymore. Will fix!

peanutfun replied:
It was written, but the test was not adapted properly. Everything works now!

chahank commented on climada/engine/impact.py:
```diff
@@ -1719,7 +1652,7 @@ def concat(cls, imp_list: Iterable, reset_event_ids: bool = False):
         - sums up the values of attributes ``eai_exp``, ``aai_exp``
         - and takes the following attributes from the first impact object in the passed
-          impact list: ``coord_exp``, ``crs``, ``unit``, ``tot_value``, ``tag``,
+          impact list: ``coord_exp``, ``crs``, ``unit``, ``tot_value``,
           ``frequency_unit``
```
```suggestion
          impact list: ``coord_exp``, ``crs``, ``unit``, ``tot_value``,
          ``frequency_unit``, ``haz_type``
```

chahank commented on climada/util/yearsets.py:
```diff
@@ -82,7 +82,6 @@ def impact_yearset(imp, sampled_years, lam=None, correction_fac=True, seed=None)
 
     #save calculations in yimp
     yimp.event_id = np.arange(1, n_sampled_years+1)
-    yimp.tag['yimp object'] = True
```
Does this potentially break a functionality? This is one of the rare instances of a possibly relevant variable to be stored in the tags object.

peanutfun replied:
Not sure 😅 All tests pass on my end, so I don't think it does? 🤷

carmensteinmann replied:
I quickly checked and don't think it would break any functionality.

chahank reviewed (changes requested):
Great progress on the quest towards removing the `Tag` from CLIMADA. Thanks!

peanutfun commented:
@chahank There is another issue: `Impact.from_eih` now has an unused argument `impfset`: https://github.com/CLIMADA-project/climada_python/pull/743/files#diff-1c276908f04d1a0ca5b1692295a586ce54b0937673711203582d3d72099ecdd0R210

The `impfset` was only used to populate the tag. I propose to deal with this the following way:
* Remove the `impfset` parameter from the method immediately
* Update all uses of `from_eih` within Climada (there are not many)
* Add a type check for the second argument. If it is an `ImpactFuncSet`, raise an error with useful error message
This way, we are *not* backward compatible but users know how to fix the issue.

chahank commented:
> @chahank There is another issue: `Impact.from_eih` now has an unused argument `impfset`: https://github.com/CLIMADA-project/climada_python/pull/743/files#diff-1c276908f04d1a0ca5b1692295a586ce54b0937673711203582d3d72099ecdd0R210
> 
> The `impfset` was only used to populate the tag. I propose to deal with this the following way:
> 
>     * Remove the `impfset` parameter from the method immediately
> 
>     * Update all uses of `from_eih` within Climada (there are not many)
> 
>     * Add a type check for the second argument. If it is an `ImpactFuncSet`, raise an error with useful error message
> 
> 
> This way, we are _not_ backward compatible but users know how to fix the issue.

Excellent idea! Although I think we do not need to make the error message. Most users will not directly use this method anyway imho.

chahank commented:
Oh, and probably you should go through the documentation and check whether the `tag` is explicitly mentioned. If yes, please remove it.

peanutfun commented:
@chahank I removed the `impfset` parameter from `Impact.from_eih` without adding an type check, as discussed. I removed the mention of Tag from the impact tutorial. It is still mentioned in other tutorials (e.g. exposures and hazard) though, but I would remove these instances once tags are removed from the respective classes.

chahank commented:
Excellent, looks good to me!

peanutfun commented:
Thanks for the input, merging now!
